### PR TITLE
[#98271560] Postgres log shipping AWS

### PR DESCRIPTION
**Pull request by saliceti**
We want the Postgres server to be able to recover from failure by shipping the logs to secure storage. Postgres uses [WAL-E](https://github.com/wal-e/wal-e) to send full backups and transaction logs to an S3 bucket.
-  We have a new playbook to configure WAL-E: [WAL-E playbook](https://github.com/alphagov/ansible-playbook-wal-e)
-  The required resources were created via terraform in [this PR](https://github.com/alphagov/tsuru-terraform/pull/76)

This particular PR adds log shipping to an AWS environment. GCE is not directly supported in WAL-E, this will be investigated in story #98365398.
# Who should review

Anyone but @saliceti, @keymon , @RichardKnop
# How to review
-  Apply to an AWS environment
-  Check log shipping is configured
-  Test backup + restore
## Check WAL-E and S3 integration

You can check that WAL-E works and is able to upload content by executing the push commands manually from the server:

``` bash
sudo su - postgres

# Base backup:
/usr/bin/envdir /etc/wal-e /usr/local/bin/wal-e --aws-instance-profile backup-push  /var/lib/postgresql/9.3/main/

# WAL push (simulate what postgres runs)
cd /var/lib/postgresql/9.3/main/
# Pick one wal file from pg_xlog/ dir
/usr/bin/envdir /etc/wal-e /usr/local/bin/wal-e --aws-instance-profile wal-push pg_xlog/00000001000000000000000B
```

Check if the configuration is done properly and files are uploaded
WAL runs from 2 places, which the playbook setups.
-  Cron job to do the "full base backup", setup to run once a week:

``` bash
$ cat /etc/cron.d/wal_e_base_backup 
#Ansible: wal_e_base_backup
0 0 * * 1 postgres /usr/bin/envdir /etc/wal-e /usr/local/bin/wal-e --aws-instance-profile backup-push  /var/lib/postgresql/9.3/main/
```
-  Postgres trigger, in `/etc/postgresql/9.3/main/postgresql.conf`. Should include:

``` bash
$ sudo cat /etc/postgresql/9.3/main/postgresql.conf | grep archive_command
archive_command = '/usr/bin/envdir /etc/wal-e /usr/local/bin/wal-e --aws-instance-profile wal-push %p'        # command to use to archive a logfile segment
```

You can check that both jobs run. For the backup job:
1. Setting up the cron job to run every 10 minutes, for example, change /etc/cron.d/wal_e_base_backup and rerun service cron restart to:

``` bash
*/10 * * * * postgres /usr/bin/envdir /etc/wal-e /usr/local/bin/wal-e --aws-instance-profile backup-push  /var/lib/postgresql/9.3/main/
```
1. Check the /var/log/syslog for the output of the wal-e upload of the base backup. Should contain `wal_e.operator.backup`: 

``` bash
wal_e.operator.backup INFO     MSG: start upload postgres version metadata#012        DETAIL: Uploading to s3://my_bucket/wal-e/basebackups_005/base_00000001000000000000000E_00000040/extended_version.txt.#012        STRUCTURED: time=2015-07-03T14:13:57.762694-00 pid=6791
```

For the wal log shipping, Postgres triggers the wal archive when gets enough data (~16MB). But we can for a timeout for testing by changing archive_timeout:
1. Change archive_timeout = 10 in /etc/postgresql/9.3/main/postgresql.conf:

``` bash
sudo sed -i 's/archive_timeout.*/archive_timeout = 10/' /etc/postgresql/9.3/main/postgresql.conf
sudo service postgresql reload
sudo -u postgres psql -c 'show archive_timeout'
```
1. Write something in postgresql.
2. Wait to see the upload in /var/log/postgresql/postgresql-9.3-main.log. Something like:

``` bash
wal_e.worker.upload INFO     MSG: begin archiving a file
    DETAIL: Uploading "pg_xlog/000000020000000000000016" to "s3://bucket/wal-e/wal_005/000000020000000000000016.lzo".
    STRUCTURED: time=2015-07-03T15:46:38.714023-00 pid=14816 action=push-wal key=s3://bucket/wal-e/wal_005/000000020000000000000016.lzo prefix=wal-e/ seg=000000020000000000000016 state=begin
wal_e.worker.upload INFO     MSG: completed archiving to a file 
    DETAIL: Archiving to "s3://bucket/wal-e/wal_005/000000020000000000000016.lzo" complete at 134.87KiB/s. 
    STRUCTURED: time=2015-07-03T15:46:39.366830-00 pid=14816 action=push-wal key=s3://bucket/wal-e/wal_005/000000020000000000000016.lzo prefix=wal-e/ rate=134.87 seg=000000020000000000000016 state=complete
```
## Check if we can restore the backup

We will test with flask application (see demo runbook as example), binding a postgresql instance:
1. Start from a environment with postgresql and WAL-E setup, with the cron job running each 10m and `archive_timeout` settings as described above.
2. Setup flask with a DB binding.
3. Add some posts to flask app (login: admin, pass: default)
4. Get a Base backup. For that wait > 10m for the cronjob to run or just login on the postgresql DB server and run the cron command manually as postgresql: `sudo -u postgres /usr/bin/envdir /etc/wal-e /usr/local/bin/wal-e --aws-instance-profile backup-push /var/lib/postgresql/9.3/main/`
5. Add some more posts in flask demo app, and wait some seconds for the DB to upload the WAL files (check logs if desired, as described before)
6. Reprovision the postgresql DB, using terraform and ansible:
    `bash
    cd tsuru-terraform/aws
    terraform taint aws_instance.postgres
    terraform apply -var env=....................... -var force-destroy=true
    cd ../../tsuru-ansible
    make aws DEPLOY_ENV=russ ARGS='-l *postgres*'
`
7. Flask now should fail because everything is broken, no DB exists, etc.
Now restore the backup, following the instructions from postgres and WAL-E:
- ssh on the postgres server
- Stop postgres and delete the data

``` bash
sudo service postgresql stop
mv /var/lib/postgresql/9.3/main /var/lib/postgresql/9.3/main.bak
mkdir /var/lib/postgresql/9.3/main
chmod 0700 "/var/lib/postgresql/9.3/main"
```
- Check the WAL-E latest base backup:

``` bash
$ sudo -u postgres /usr/bin/envdir /etc/wal-e /usr/local/bin/wal-e --aws-instance-profile backup-list
name    last_modified    expanded_size_bytes    wal_segment_backup_start    wal_segment_offset_backup_start    wal_segment_backup_stop    wal_segment_offset_backup_stop
base_00000001000000000000000A_00000040    2015-07-02T16:05:09.000Z        00000001000000000000000A    00000040        
base_00000001000000000000000E_00000040    2015-07-03T14:14:01.000Z        00000001000000000000000E    00000040
```
- Pull the latest one:

``` bash
/usr/bin/envdir /etc/wal-e /usr/local/bin/wal-e --aws-instance-profile backup-fetch /var/lib/postgresql/9.3/main/ base_00000001000000000000000A_00000040
```
- Create a file `/var/lib/postgresql/9.3/main/recovery.conf` to tell postgresql to trigger recovery and how to pull the WAL files:

``` bash
cat <<"EOF"  | sudo -u postgres tee /var/lib/postgresql/9.3/main/recovery.conf
restore_command ='/usr/bin/envdir /etc/wal-e /usr/local/bin/wal-e --aws-instance-profile wal-fetch "%f" "%p"'
EOF
```
- Start postgres: sudo service start postgresql it should take some time and recover the DB.
- Refresh flask app (retry some times because the DB pool might have dead connections)

### CODE REVIEW

**Comment by RichardKnop** [comment]
:+1:

**Comment by dcarley** [comment]
> Pull the latest one:
> /usr/bin/envdir /etc/wal-e /usr/local/bin/wal-e --aws-instance-profile backup-fetch /var/lib/postgresql

I couldn't get this to work. It gave me an exception and then hung, which after waiting a long time I cancelled:

```
ubuntu@ip-10-128-11-86:~$ sudo -u postgres /usr/bin/envdir /etc/wal-e /usr/local/bin/wal-e --aws-instance-profile backup-list
name    last_modified   expanded_size_bytes     wal_segment_backup_start        wal_segment_offset_backup_start wal_segment_backup_stop wal_segment_offset_backup_stop
base_000000010000000000000002_00000040  2015-07-09T13:11:22.000Z                000000010000000000000002        00000040
ubuntu@ip-10-128-11-86:~$ sudo -u postgres /usr/bin/envdir /etc/wal-e /usr/local/bin/wal-e --aws-instance-profile backup-fetch /var/lib/postgresql/9.3/main/ base_000000010000000000000002_00000040
wal_e.worker.s3.s3_worker INFO     MSG: beginning partition download
        DETAIL: The partition being downloaded is part_00000000.tar.lzo.
        HINT: The absolute S3 key is wal-e/basebackups_005/base_000000010000000000000002_00000040/tar_partitions/part_00000000.tar.lzo.
        STRUCTURED: time=2015-07-09T13:48:42.642117-00 pid=29142
Traceback (most recent call last):
    result = self._run(*self.args, **self.kwargs)
  File "/usr/local/lib/python2.7/dist-packages/wal_e/blobstore/s3/s3_util.py", line 154, in write_and_return_error
    stream.close()
  File "/usr/local/lib/python2.7/dist-packages/wal_e/pipebuf.py", line 261, in close
    del self._bd
AttributeError: _bd
<Greenlet at 0x7f8e8165eaf0: write_and_return_error(<Key: dcarley-mcp.postgres.backup,wal-e/basebackup, <wal_e.pipebuf.NonBlockBufferedWriter object at 0x)> failed with AttributeError




^CKeyboardInterrupt
Traceback (most recent call last):
  File "/usr/local/bin/wal-e", line 9, in <module>
    load_entry_point('wal-e==0.7.0', 'console_scripts', 'wal-e')()
  File "/usr/local/lib/python2.7/dist-packages/wal_e/cmd.py", line 516, in main
    pool_size=args.pool_size)
  File "/usr/local/lib/python2.7/dist-packages/wal_e/operator/backup.py", line 153, in database_fetch
    p.join(raise_error=True)
  File "/usr/local/lib/python2.7/dist-packages/gevent/pool.py", line 95, in join
    self._empty_event.wait(timeout=timeout)
  File "/usr/local/lib/python2.7/dist-packages/gevent/event.py", line 77, in wait
    result = self.hub.switch()
  File "/usr/local/lib/python2.7/dist-packages/gevent/hub.py", line 338, in switch
    return greenlet.switch(self)
KeyboardInterrupt
```

NB: I only did one base backup but I don't think that matters.

File: site-aws.yml [hunk 1]
```diff
@@ -11,6 +11,7 @@
     storage_type: "s3"
     s3_region: "eu-west-1"
     s3_bucket: "{{ deploy_env }}-mcp.registry.storage"
+    postgres_backup_s3_bucket: "{{ deploy_env }}-mcp.postgres.backup"
```
Comment (dcarley) [inline site-aws.yml]: I think this is being passed to the registry playbook and not being used?

Comment (keymon) [inline site-aws.yml]: You are right, I will remove that.

**Comment by keymon** [comment]
The exception you are facing is interesting, we did not have the same problem before. It seems to be [a bug](https://github.com/wal-e/wal-e/commit/e3fd6be0ef1109ca5b15f7b6540a2bf773e0dc34) solved in the [version 0.8.0 of WAL-E](https://github.com/wal-e/wal-e/compare/f3edb82baba15450933a2501005a4a4b261a72d8...master#diff-e77e4d3d46d72b3cec4da954c4597bbaL261) 

For the GCE story, they did a new [PR of the playbook (merged already)](https://github.com/alphagov/ansible-playbook-wal-e/pull/2) of wal-e which [uses version 0.8.0 of WAL-E](https://github.com/alphagov/ansible-playbook-wal-e/commit/554f25998a3affc819a5cacc990404571dc4add7), but there is no release done.

I will:
- Create a new release in ansible-playbook-wal-e
- Use the new version here

**Comment by dcarley** [comment]
Should we be changing the value of `archive_timeout` and the frequency for the base backup cronjob to lower values? Are the defaults sufficient for us now?

**Comment by keymon** [comment]
That is a quite good question, but not sure if should be answered in this story. We were just implementing the log shipping, not solving the final backup configuration.

It is possible to override all those values, though.

**Comment by dcarley** [comment]
I guess it depends how we intend to use it right now. I was thinking, in the context of this story, for a demo environment we may not write enough data to ever trigger the 16MB threshold and the environment might not have been up for a week, so we may never see any logs shipped.

**Comment by keymon** [comment]
In that case I propose, for our test environments, backups in cron every hour and archive_timeout=60 (1 minute).

**Comment by russss** [comment]
Those sound decent for the test environment.

These parameters aren't crucial if there's a slave database running (which is the next step to implement) - in the presence of a slave, the archived log files are only necessary if you want to bring up a new slave, or do a point-in-time recovery. 

If there's not a slave and you're relying on archived logs only for backup, the tuning for these parameters is a bit more important. But in that case you're probably going to end up with an unacceptable amount of data loss (and downtime) on failure.

**Comment by dcarley** [comment]
Thanks @russss 

Just had a discussion with @keymon and Brett in the office. We have a story to add a slave and we probably need to think more carefully about our use cases before we come up with new numbers, so we'll stick with the defaults for now.

**Comment by dcarley** [comment]
Hector's commits have fixed the exception, but it's still hanging for me:

```
ubuntu@ip-10-128-11-235:~$ sudo -u postgres /usr/bin/envdir /etc/wal-e /usr/local/bin/wal-e --aws-instance-profile backup-fetch /var/lib/postgresql/9.3/main/ base_000000010000000000000002_00000040
wal_e.main   INFO     MSG: starting WAL-E
        DETAIL: The subcommand is "backup-fetch".
        STRUCTURED: time=2015-07-09T15:34:04.544810-00 pid=29125
wal_e.worker.s3.s3_worker INFO     MSG: beginning partition download
        DETAIL: The partition being downloaded is part_00000000.tar.lzo.
        HINT: The absolute S3 key is wal-e/basebackups_005/base_000000010000000000000002_00000040/tar_partitions/part_00000000.tar.lzo.
        STRUCTURED: time=2015-07-09T15:34:05.350390-00 pid=29125





^CKeyboardInterrupt
Traceback (most recent call last):
  File "/usr/local/bin/wal-e", line 9, in <module>
    load_entry_point('wal-e==0.8.0', 'console_scripts', 'wal-e')()
  File "/usr/local/lib/python2.7/dist-packages/wal_e/cmd.py", line 545, in main
    pool_size=args.pool_size)
  File "/usr/local/lib/python2.7/dist-packages/wal_e/operator/backup.py", line 156, in database_fetch
    p.join(raise_error=True)
  File "/usr/local/lib/python2.7/dist-packages/gevent/pool.py", line 95, in join
    self._empty_event.wait(timeout=timeout)
  File "/usr/local/lib/python2.7/dist-packages/gevent/event.py", line 77, in wait
    result = self.hub.switch()
  File "/usr/local/lib/python2.7/dist-packages/gevent/hub.py", line 338, in switch
    return greenlet.switch(self)
KeyboardInterrupt
```

I'm going to try again with a fresh instance and bucket tomorrow morning.

**Comment by keymon** [comment]
@dcarley before you retry recreating the bucket, can you also try:
- Change the path to /tmp/test: `sudo -u postgres /usr/bin/envdir /etc/wal-e /usr/local/bin/wal-e --aws-instance-profile backup-fetch /tmp/test/ base_000000010000000000000002_00000040
  `. If that works:
- Try again, but be sure that the postgresql server is stopped first.

**Comment by dcarley** [comment]
Darn, it was permissions related. After `sudo mkdir /var/lib/postgresql/9.3/main` and `chmod 0700 "/var/lib/postgresql/9.3/main"` the directory was owned by `root:root` and `sudo -u postgres … wal-e` wasn't able to write to it. If I get a chance and find the appropriate place then I may file a bug.

I also wasn't able to get the Flask app running against because the IP of the postgres server has changed and the service entry in Tsuru needs updating, per the note in #110. However I was able to see the data using `psql`.

**Pull request merged**